CDITests.testCDIInjectContexts

The relevant specification version and section number(s): 2.1

The coordinates of the challenged test(s): com.ibm.jbatch.tck.tests.jslxml.CDITests.testCDIInjectContexts

The exact TCK version: 2.1

The implementation being tested, including name and company: JBeret, by Red Hat, Inc.

A full description of why the test is invalid and what the correct behavior is believed to be:
Any supporting material; debug logs, test output, test logs, run scripts, etc.:


I noticed 2 issues with TCK test [CDITests.testCDIInjectContexts](https://github.com/eclipse-ee4j/batch-tck/blob/master/com.ibm.jbatch.tck/src/main/java/com/ibm/jbatch/tck/tests/jslxml/CDITests.java#L86):
 
1, this test uses a batchlet [DependentScopedBatchletContexts](https://github.com/eclipse-ee4j/batch-tck/blob/master/com.ibm.jbatch.tck/src/main/java/com/ibm/jbatch/tck/artifacts/cdi/DependentScopedBatchletContexts.java#L35)  which injects JobContext and StepContext in various ways (field, constructor, and method injections). But all fields are already annotated with @Inject, which means they are already field-injected. So even if one implementation does not support method or constructor injection, the target field already contains the expected value. So we should tighten this test by removing the @Inject from those fields receiving constructor or method injections.
 
2, step exit status from process() method return value, or StepContext.setExitStatus(...)? 
 
A batchlet step can get its exit status from either process() method return value, or StepContext.setExitStatus(), but which takes precedence? The test assumes setExitStatus(...) wins and ignore the return value. I couldn't find a clear language in the spec.
 
One may argue that setExitStatus() call is more explicit so it should win. But return value from process() is equally strong in this regard, since this is exactly the purpose of this return value, and no other use of this return value exists.
 
I suggest we change the test such that the batchlet process method returns the same value as the value in setExitStatus(), and the test always receives the consistent exit status.
```
[ERROR] Failures:
[ERROR]   CDITests.testCDIInjectContexts:116 Test fails - unexpected step exit status ==> expected: <13:13:13:> but was: <OK>
[ERROR]   CDITests.testCDIInjectContexts:116 Test fails - unexpected step exit status ==> expected: <14:14:14:> but was: <OK>
[ERROR]   CDITests.testCDIInjectContexts:116 Test fails - unexpected step exit status ==> expected: <15:15:15:> but was: <OK>
```

--- a/com.ibm.jbatch.tck/src/main/java/com/ibm/jbatch/tck/artifacts/cdi/DependentScopedBatchletContexts.java
+++ b/com.ibm.jbatch.tck/src/main/java/com/ibm/jbatch/tck/artifacts/cdi/DependentScopedBatchletContexts.java
@@ -36,12 +36,18 @@ import jakarta.inject.Named;
 @Named("CDIDependentScopedBatchletContexts")
 public class DependentScopedBatchletContexts implements Batchlet {
 
+    //
+    // As mentioned in issue: https://github.com/eclipse-ee4j/batch-tck/issues/49
+    // the fields getting set via constructor/method parameter injection should not
+    // themselves be annotated with "@Inject".
+    //
     @Inject JobContext jf; 
-    @Inject JobContext jc; 
-    @Inject JobContext jm; 
+    JobContext jc;
+    JobContext jm;
+
     @Inject StepContext sf; 
-    @Inject StepContext sc; 
-    @Inject StepContext sm; 
+    StepContext sc;
+    StepContext sm;
 
     @Inject
     DependentScopedBatchletContexts(JobContext jc, StepContext sc) {
@@ -61,13 +67,21 @@ public class DependentScopedBatchletContexts implements Batchlet {
 
     @Override
     public String process() throws Exception {
+
         updateJobExitStatus(jf);
         updateJobExitStatus(jc);
         updateJobExitStatus(jm);
         updateStepExitStatus(sf);
         updateStepExitStatus(sc);
         updateStepExitStatus(sm);
-        return "OK";
+
+        //
+        // Spec doesn't clearly specify precedence between exit status set into StepContext vs. exit status returned by Batchlet process().
+        // To avoid ambiguity, ensure the same value is used by both mechanisms.
+        //
+        // See: https://github.com/eclipse-ee4j/batch-tck/issues/49
+        //
+        return sm.getExitStatus();
     }
 
     private void updateJobExitStatus(JobContext jobCtx) {

